File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        letterMap = {
            "2" : "abc",
            "3" : "def",
            "4" : "ghi",
            "5" : "jkl",
            "6" : "mno",
            "7" : "pqrs",
            "8" : "tuv",
            "9" : "wxyz"
            }

        chars = [letterMap[d] for d in digits]
        return [''.join(i) for i in list(itertools.product(*chars))] if digits else []

File: 17-letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.java
class Solution {
    String[] dict = new String[] {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        if (digits.equals("")) return res;

        helper(new StringBuilder(), 0, digits, res);

        return res;
    }

    private void helper(StringBuilder sb, int ind, String digits, List<String> res) {
        if (digits.length() == ind) {
            res.add(sb.toString());
            return;
        }

        for (char c: dict[digits.charAt(ind) - '0'].toCharArray()) {
            sb.append(c);
            helper(sb, ind + 1, digits, res);
            sb.setLength(sb.length() - 1);
        }
    }
}